// docsify-tools/build/api-documenter/cli/YamlAction.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
const BaseAction_1 = require("./BaseAction");
const YamlDocumenter_1 = require("../documenters/YamlDocumenter");
const OfficeYamlDocumenter_1 = require("../documenters/OfficeYamlDocumenter");
class YamlAction extends BaseAction_1.BaseAction {
    constructor(parser) {
        super({
            actionName: 'yaml',
            summary: 'Generate documentation as universal reference YAML files (*.yml)',
            documentation: 'Generates API documentation as a collection of files conforming'
                + ' to the universal reference YAML format, which is used by the docs.microsoft.com'
                + ' pipeline.'
        });
    }
    onDefineParameters() {
        super.onDefineParameters();
        this._officeParameter = this.defineFlagParameter({
            parameterLongName: '--office',
            description: `Enables some additional features specific to Office Add-ins`
        });
    }
    onExecute() {
        const apiModel = this.buildApiModel();
        const yamlDocumenter = this._officeParameter.value
            ? new OfficeYamlDocumenter_1.OfficeYamlDocumenter(apiModel, this.inputFolder)
            : new YamlDocumenter_1.YamlDocumenter(apiModel);
        yamlDocumenter.generateFiles(this.outputFolder);
        return Promise.resolve();
    }
}
exports.YamlAction = YamlAction;
